process/block/metablock.go

```diff
@@ -1205,9 +1205,68 @@ func (mp *metaProcessor) updateState(lastMetaBlock data.HeaderHandler) {
 	)
 }
 
-func (mp *metaProcessor) getLastSelfNotarizedHeaderByShard(_ uint32) (data.HeaderHandler, []byte) {
-	//TODO: Implement mechanism to extract last meta header notarized by the given shard if this info will be needed later
-	return nil, nil
+func (mp *metaProcessor) getLastSelfNotarizedHeaderByShard(
+	metaBlock *block.MetaBlock,
+	shardID uint32,
+) (data.HeaderHandler, []byte) {
+	maxNotarizedNonce := uint64(0)
+	lastNotarizedMetaHeader, lastNotarizedMetaHeaderHash, err := mp.blockTracker.GetLastSelfNotarizedHeader(shardID)
+	if err != nil {
+		log.Warn("getLastSelfNotarizedHeaderByShard.GetLastSelfNotarizedHeader",
+			"shard", shardID,
+			"error", err.Error())
+		return nil, nil
+	}
+
+	for _, shardData := range metaBlock.ShardInfo {
+		if shardData.ShardID != shardID {
+			continue
+		}
+
+		mp.hdrsForCurrBlock.mutHdrsForBlock.RLock()
+		headerInfo, ok := mp.hdrsForCurrBlock.hdrHashAndInfo[string(shardData.HeaderHash)]
+		mp.hdrsForCurrBlock.mutHdrsForBlock.RUnlock()
+		if !ok {
+			log.Debug("getLastSelfNotarizedHeaderByShard",
+				"error", process.ErrMissingHeader,
+				"hash", shardData.HeaderHash)
+			continue
+		}
+
+		shardHeader, ok := headerInfo.hdr.(*block.Header)
+		if !ok {
+			log.Debug("getLastSelfNotarizedHeaderByShard",
+				"error", process.ErrWrongTypeAssertion,
+				"hash", shardData.HeaderHash)
+			continue
+		}
+
+		for _, metaHash := range shardHeader.MetaBlockHashes {
+			metaHeader, err := process.GetMetaHeader(metaHash, mp.dataPool.Headers(), mp.marshalizer, mp.store)
```

Comment:
I think having LastIncludedMetaNonce already in the previous metablock (the one from blockchain) + the information in the current metablock, the new LastIncludedMetaNonce for every shard can be calculated without reading it from storage. Effectively it would be LastIncludedMetaNonce + len(shardHeader.MetaBlockHashes) - 1 if the last included metahash from the previous shardHeader is equal with the first of the current one.